pkg/p2p/internal/send_chan.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	cerror "github.com/pingcap/tiflow/pkg/errors"
+	proto "github.com/pingcap/tiflow/proto/p2p"
+	"go.uber.org/zap"
+)
+
+// SendChan is a specialized channel used to implement
+// the asynchronous interface of the MessageClient.
+// SendChan is a MPSC channel.
+type SendChan struct {
+	mu      sync.Mutex
+	buf     []*proto.MessageEntry
+	sendIdx int64
+	recvIdx int64
+
+	// hasNewMsg is a buffered channel that the sending goroutine(s)
+	// use to notify the receiver end.
+	hasNewMsg chan struct{}
+
+	// notifyChan is an unbuffered channel that
+	// SendSync writes to in order to determine whether the
+	// consumer is ready.
+	notifyChan chan struct{}
+
+	cap int64
+}
+
+// NewSendChan returns a new SendChan.
+func NewSendChan(cap int64) *SendChan {
+	return &SendChan{
+		buf:        make([]*proto.MessageEntry, cap),
+		hasNewMsg:  make(chan struct{}, 1),
+		notifyChan: make(chan struct{}),
+		cap:        cap,
+	}
+}
+
+// SendSync sends a message synchronously.
+// The performance of SendSync is about 500x slower than
+// SendAsync when CPU is saturated, so please refrain from
+// using it unless necessary or for testing.
+func (c *SendChan) SendSync(
+	ctx context.Context,
+	topic string,
+	value []byte,
+	closeCh <-chan struct{},
+	nextSeq func() int64,
+) (int64, error) {
+	ticker := time.NewTicker(10 * time.Millisecond)
```

Comment:
make 10ms configurable? Benchmark performance is related to it.

Reply:
`SendSync` is a marginal use case. Not sure if it is worth adding a config?

Reply:
DM in dataflow engine mostly use blocking sending

https://github.com/hanfei1991/microcosm/blob/3ed20d9ee8677fb1d52af5fc3c2c15c00c19c144/pkg/dm/message_pair.go#L52

but we can add the configuration when problem really happens.